src/context.jl

```diff
@@ -119,6 +179,11 @@ struct Context
     debug::Bool
     tempfile::Union{String, Nothing}
     streaming::Bool
+    types::Union{Nothing, Type, AbstractVector, AbstractDict, Function}
```

Comment:
the least nice thing about this PR, in my view, is that (i) it requires yet-another field in `Context` and (ii) one with this long-winded type `Union{Nothing, Type, AbstractVector, AbstractDict, Function}`

On (i), i couldn't think of another way to pass this info through to the parsing context... it's kinda what `Context` is for.

On (ii), i considered whether when constructing the `Context` we could construct a new `types` with a fixed type e.g. `types::Vector`, but i think that actually becomes more tricky, since we always want the same flexibility with `types` whether we're using it _before_ parsing (like in Context) or _during_ parsing (like when we hit an extra column), so i decided against this.

Thoughts welcome!

Reply:
This is probably fine. Still reviewing the rest of the PR